### PR TITLE
bpo-39924 optimize the pathlib.readlink behavior when the readlink is not existed in os module

### PR DESCRIPTION
**Pull request by Zheaoli**
https://bugs.python.org/issue39924

https://bugs.python.org/issue39924

### CODE REVIEW

File: Lib/pathlib.py [hunk 1]
```diff
@@ -445,6 +445,8 @@ def symlink(a, b, target_is_directory):
 
     # Helper for resolve()
```
Comment (barneygale) [inline Lib/pathlib.py:446]: I'd suggest implementing this like `lchmod` above, i.e.:

```python
if hasattr(os, "readlink"):
    readlink = os.readlink
else:
    def readlink(self, path):
        raise NotImplementedError("readlink() not available on this system")
```

This way you only do the `hasattr` check once.

Comment (Zheaoli) [reply Lib/pathlib.py:446]: Right, it makes sense

File: Lib/pathlib.py [hunk 2]
```diff
@@ -444,8 +444,12 @@ def symlink(a, b, target_is_directory):
     utime = os.utime
 
     # Helper for resolve()
-    def readlink(self, path):
-        return os.readlink(path)
+    if hasattr(os, "readlink"):
+        def readlink(self, path):
+            return os.readlink(path)
```
Comment (barneygale) [inline Lib/pathlib.py]: `readlink = os.readlink`

**Comment by barneygale** [comment]
Would be good to address this part of the bug report also:

> `symlink()` doesn't need to guard against `os.symlink()` not accepting `target_is_directory` on non-Windows platforms; this has been fixed since 3.3

It's essentially the exact same code again (`if hasattr` etc), but for `os.symlink` rather than `os.readlink`.

File: Lib/pathlib.py (hunk 2 above)
Comment (Zheaoli) [reply Lib/pathlib.py]: actually, it will make the `readlink` from a method to a static method. In other words, it will change the API behavior

**Comment by Zheaoli** [comment]
> Would be good to address this part of the bug report also:

what's your point? I don't get it

File: Lib/pathlib.py (hunk 2 above)
Comment (barneygale) [reply Lib/pathlib.py]: How does that change API behaviour?

Comment (Zheaoli) [reply Lib/pathlib.py]: from a method to a static method.

Although people can use it as usual, it changes the API implementation detail. I think it should be an API behavior-change.

Comment (barneygale) [reply Lib/pathlib.py]: Can you show a repro case? I still can't see how this would break an API.

Comment (Zheaoli) [reply Lib/pathlib.py]: it will not break the API, It will change the method type. I can't get a better word to describe it. :(

Comment (barneygale) [reply Lib/pathlib.py]: I get that, but I don't see why it matters. Everything else in `_NormalAccessor` uses `os` functions directly without wrapping.

Comment (Zheaoli) [reply Lib/pathlib.py]: it makes sense, I will update the code

**Comment by iritkatriel** [comment]
https://bugs.python.org/issue39924 is closed. What is the status of this PR?

**Comment by barneygale** [comment]
As b57e045320d1d2a70eab236b7d31a3ebb75037c3 has landed, I believe this PR could be closed.

**Comment by iritkatriel** [comment]
Thank you.

**Pull request closed without merging**